### PR TITLE
[Merged by Bors] - feat: semisimple rings are Artinian

### PR DESCRIPTION
**Pull request by alreadydone**
The main result `IsSemisimpleModule.finite_tfae` shows that for a semisimple module, `Module.Finite R M ↔ IsNoetherian R M ↔ IsArtinian R M ↔ IsFiniteLength R M ↔` direct sum decomposition into finitely many simple modules (F25 in Chapter 28 of Lorenz's *Algebra II*). The instance from IsSemisimpleModule + Module.Finite to IsNoetherian is proven before and used in the proof, while the instance to IsArtinian is added after the proof. As a consequence, semisimple rings are Artinian.

Also add easy lemmas `isNoetherian_range`, `LinearEquiv.isNoetherian_iff`, `isNoetherian_sup`, `isNoetherian_iSup` (and Artinian counterparts) and `IsSemisimpleModule.exists_setIndependent_sSup_simples_eq_top`.

---

- [x] depends on: #17525
- [x] depends on: #17478 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/FiniteLength.lean [hunk 1]
```diff
@@ -76,3 +75,28 @@ theorem isFiniteLength_iff_exists_compositionSeries :
...
+    have := fin.to_subtype
+    exact ⟨isNoetherian_iSup, isArtinian_iSup⟩
+  tfae_have 4 → 2 := And.left
+  tfae_have 4 → 3 := And.right
+  tfae_finish
+
+instance [IsSemisimpleModule R M] [Module.Finite R M] : IsArtinian R M :=
+  (IsSemisimpleModule.finite_tfae.out 0 2).mp ‹_›
+
+example [IsSemisimpleRing R] : IsNoetherianRing R := inferInstance
```
Comment (mattrobball) [inline Mathlib/RingTheory/FiniteLength.lean:101]: These should be tests importing Mathlib if you want to assured that downstream users can synthesize them. 

If you want to document to the user that they can be found, you can record a comment.

Comment (alreadydone) [reply Mathlib/RingTheory/FiniteLength.lean:101]: Where are these tests usually located? I think these are key results that follow from instance just proven so they naturally fit here, and I would hesitate to move them to another file. (The Noetherian example is already available in the SimpleModule file, but I put it here for coherence.) I can definitely make them comments, but I think making them examples ensures that no one (doing refactors etc.) would accidentally break them.

Comment (mattrobball) [reply Mathlib/RingTheory/FiniteLength.lean:101]: Working in this file doesn't guarantee that they work downstream of the file. To avoid future breakage, these should be in a file in the `tests` folder with the `inferInstance`s and `import Mathlib`. This assures that if someone changes any part of the library and breaks the current synthesis path they will be responsible for restoring it.

Comment (alreadydone) [reply Mathlib/RingTheory/FiniteLength.lean:101]: Let me make them comments then. They'll probably break only when the reducibility of IsNoetherianRing/IsArtinianRing, but in that case I'd expect a bunch of other things would also break.

File: Mathlib/RingTheory/FiniteLength.lean [hunk 2]
```diff
@@ -76,3 +75,28 @@ theorem isFiniteLength_iff_exists_compositionSeries :
...
+example [IsSemisimpleRing R] : IsArtinianRing R := inferInstance
```
Comment (alreadydone) [inline Mathlib/RingTheory/FiniteLength.lean:102]:
```suggestion
/- The following instances are now automatic: 
example [IsSemisimpleRing R] : IsNoetherianRing R := inferInstance
example [IsSemisimpleRing R] : IsArtinianRing R := inferInstance
-/
```

File: Mathlib/RingTheory/FiniteLength.lean [hunk 3]
```diff
@@ -76,3 +75,30 @@ theorem isFiniteLength_iff_exists_compositionSeries :
...
+    have := fin.to_subtype
+    exact ⟨isNoetherian_iSup, isArtinian_iSup⟩
+  tfae_have 4 → 2 := And.left
+  tfae_have 4 → 3 := And.right
+  tfae_finish
+
+instance [IsSemisimpleModule R M] [Module.Finite R M] : IsArtinian R M :=
+  (IsSemisimpleModule.finite_tfae.out 0 2).mp ‹_›
+
+/- The following instances are now automatic:
```
Comment (alreadydone) [inline Mathlib/RingTheory/FiniteLength.lean:101]:
```suggestion
/- The following instances are now automatic:
```

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Comment by mattrobball** [comment]
I'm not sure why there is not a test file @jcommelin

**Comment by mattrobball** [comment]
I'd like to see a plan for one first

bors r-

**Comment by mattrobball** [comment]
Yes, there are existing uses as such but we have already seen that often hide undesirable behavior and are often poorly motivated. I think they should all be moved/removed (and some have) but that is out of scope here. Inclusion of new `example`s should meet a higher bar for justification. 

My rationale: Most of the `example`s I've seen are to demonstrate that instance synthesis succeeds. However, success is highly dependent on the imports. For the end user, who most likely will be working with additional imports, `example`s declared directly after the relevant instance declaration in the original file will not serve as a guarantee their code works as expected. For mathematically well-motived instances (as this one definitely is), we want to lock in that success of type class search throughout the whole library. This can only be done when all available instances are in scope. 

Additionally, since `example`s add nothing to the environment they confuse anything, like a linter, that uses the environment. We don't want to keep adding to `noshake.json` for each of them, for example. (This is not the case here but still.) 

TLDR: yes these instances searches are of sufficient importance that we want to make sure they keep working indefinitely. We also want to model best practices going forward.

**Comment by mattrobball** [comment]
I appreciate your patience with me. I think the request is important enough relative to the effort required. Please let me know if there is anything I can do to help.

**Comment by alreadydone** [comment]
Thanks for the explanation. I'll add the tests in test/instances. Though I still have some doubts: it's possible that when importing the whole Mathlib, a declared instance itself may fail to synthesize (it might timeout, or fail due to some bad non-defeq instance introduced in between). Does it mean we should add every instance to the tests? Or is CI already checking every instance can be synthesized after importing Mathlib, like the simpNF linter is doing to every simp lemma?

File: test/instances/Ring_finiteness.lean [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+variable (R : Type*) [Ring R] [IsSemisimpleRing R]
+
+/-- info: isNoetherian_of_isNoetherianRing_of_finite R R -/
+#guard_msgs in
+#synth IsNoetherianRing R
+
+/-- info: instIsArtinianOfIsSemisimpleModuleOfFinite -/
+#guard_msgs in
+#synth IsArtinianRing R
```
Comment (alreadydone) [inline test/instances/Ring_finiteness.lean:13]: Okay I've added the tests. The first result is not `IsSemisimpleModule.instIsNoetherianOfFinite` unlike the second result, probably because [that instance](https://github.com/leanprover-community/mathlib4/pull/17557/files#diff-95d7011cee766c6bd49b18c53feddf48cac70fc5b1a206eafa18d54b5bb6f5c0R226-R228) has low priority. I made the priority low because there's also an unconditional instance from `IsNoetherian` to `Module.Finite`.

Comment (mattrobball) [reply test/instances/Ring_finiteness.lean:13]: Oh! But I was asking for 
```lean
example : IsNoetherianRing := inferInstance 
example : IsArtinianRing := inferInstance 
```
The particular synthesis path is not as important as the fact that end users can synthesize the Artinian and Noetherian classes. 

Apologies for being unclear.

**Review by mattrobball — APPROVED** [review]

File: test/instances/Ring_finiteness.lean (hunk 4 above)
Comment (alreadydone) [reply test/instances/Ring_finiteness.lean:13]: I agree this is better! I was just copying from the other file and adjust.

Comment (alreadydone) [inline test/instances/Ring_finiteness.lean:13]:
```suggestion
example : IsNoetherianRing R := inferInstance
example : IsArtinianRing R := inferInstance
```

**Comment by mattrobball** [comment]
Thanks! 

bors merge
